### PR TITLE
Expand the list of attention implementations compatible with packing

### PR DESCRIPTION
**Pull request by mariosasko**
# What does this PR do?

Expands the list of attention implementations compatible with packing in `SFTTrainer` to include `eager`, `sdpa`, `flex_attention`, and `flash_attention_4`, and updates the warnings accordingly.

PS: Support for the packing/padding-free case for `eager`/`sdpa`/`flex_attention` was added in  https://github.com/huggingface/transformers/pull/39194

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Low risk: changes only adjust which `attn_implementation` values are treated as compatible for padding-free/packing warnings and remove an unused constant in `DPOTrainer`. No training logic or data handling paths are otherwise modified.
> 
> **Overview**
> Expands `SFTTrainer`'s set of attention implementations considered compatible with *padding-free/packing* to include `eager`, `sdpa`, `flex_attention`, and `flash_attention_4`, and updates the related compatibility checks.
> 
> Reworks the warning messages emitted when `padding_free` or BFD packing is enabled with an incompatible attention implementation to show a clearer, bullet-listed set of supported variants. Also removes the unused `FLASH_ATTENTION_VARIANTS` constant from `dpo_trainer.py`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 02e073262bda6f71e9a5aef6afccb04dcf6ec94e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
I'm not entirely sure it's properly supported yet:

```python
import torch
from transformers import AutoModelForCausalLM

MODEL_NAME = "HuggingFaceTB/SmolLM2-135M"


def run_test(attn_implementation: str, seqs: list[torch.Tensor]) -> bool:
    print(f"\n{'='*55}")
    print(f"  attn_implementation = {attn_implementation!r}")
    print(f"{'='*55}")

    model = AutoModelForCausalLM.from_pretrained(
        MODEL_NAME, attn_implementation=attn_implementation, dtype=torch.bfloat16, device_map="auto"
    )
    model = model.eval()

    with torch.no_grad():
        # Reference: each sequence independently
        refs = [model(input_ids=s).logits for s in seqs]

        # Padding-free: all sequences flattened into one
        seq_lengths = [s.shape[1] for s in seqs]
        flat_ids = torch.cat(seqs, dim=1)  # (1, total_len)
        position_ids = torch.cat([torch.arange(L, device="cuda") for L in seq_lengths]).unsqueeze(0)
        flat_out = model(input_ids=flat_ids, position_ids=position_ids).logits

        # Compare split outputs against individual references
        splits = flat_out.split(seq_lengths, dim=1)
        all_pass = True
        for i, (ref, pf) in enumerate(zip(refs, splits)):
            ok = torch.allclose(ref, pf, atol=1e-4)
            diff = (ref - pf).abs().max().item()
            print(f"  seq[{i}] (len={seq_lengths[i]}): {'✅' if ok else '❌'}  max_diff={diff:.2e}")
            all_pass = all_pass and ok


if __name__ == "__main__":
    torch.manual_seed(42)
    seqs = [
        torch.randint(100, 1000, (1, 6), device="cuda"),
        torch.randint(100, 1000, (1, 4), device="cuda"),
        torch.randint(100, 1000, (1, 7), device="cuda"),
    ]
    for attn_impl in ["eager", "sdpa", "flex_attention", "kernels-community/flash-attn2"]:
        run_test(attn_impl, seqs)
```

```

=======================================================
  attn_implementation = 'eager'
=======================================================
Loading weights: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 6994.41it/s]
  seq[0] (len=6): ❌  max_diff=5.00e-01
  seq[1] (len=4): ❌  max_diff=1.26e+01
  seq[2] (len=7): ❌  max_diff=1.42e+01

=======================================================
  attn_implementation = 'sdpa'
=======================================================
Loading weights: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 7099.13it/s]
  seq[0] (len=6): ✅  max_diff=0.00e+00
  seq[1] (len=4): ❌  max_diff=1.25e+01
  seq[2] (len=7): ❌  max_diff=1.41e+01

=======================================================
  attn_implementation = 'flex_attention'
=======================================================
Loading weights: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 7859.75it/s]
/fsx/qgallouedec/miniconda3/envs/trl/lib/python3.13/site-packages/torch/nn/attention/flex_attention.py:1559: FutureWarning: return_lse is deprecated and will be removed in v2.10. Please use return_aux=AuxRequest(lse=True) instead.
  _warn_once(
  seq[0] (len=6): ✅  max_diff=0.00e+00
  seq[1] (len=4): ❌  max_diff=1.25e+01
  seq[2] (len=7): ❌  max_diff=1.41e+01

=======================================================
  attn_implementation = 'kernels-community/flash-attn2'
=======================================================
Fetching 22 files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 22/22 [00:00<00:00, 5511.24it/s]
Download complete: : 0.00B [00:00, ?B/s]                                                                                            | 0/22 [00:00<?, ?it/s]
Loading weights: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 7735.26it/s]
  seq[0] (len=6): ✅  max_diff=0.00e+00
  seq[1] (len=4): ✅  max_diff=0.00e+00
  seq[2] (len=7): ✅  max_diff=0.00e+00
```

**Comment by dvdimitrov13** [comment]
Hi @mariosasko, @qgallouedec — running into exactly this on Blackwell (sm_120, no flash-attn wheel available) while training Gemma-4-E4B-it. Quentin's test result replicates perfectly, and I have a `PackedSFTCollator` that might be the missing piece.

**Replicating your test locally** (CPU, SmolLM2-135M, `attn_implementation="sdpa"`, fp32, same 3 sequences of lengths 6/4/7):

| Variant | seq[0] (len=6) | seq[1] (len=4) | seq[2] (len=7) |
|---|---|---|---|
| No mask, position_ids only (@qgallouedec's test) | 4.6e-5 ✅ | **1.14e+01** ❌ | **9.18e+00** ❌ |
| 2D pad mask only | first-in-pack ✅ | **1.55e+01** ❌ | **1.30e+01** ❌ |
| **4D block-diag mask + position_ids reset** | **4.7e-5** ✅ | **3.6e-5** ✅ | **5.3e-5** ✅ |

So the transformers side is fine — stock models DO respect a 4D block-diag mask (#39194 did its job). The gap is that nothing in TRL's bfd path actually *emits* that mask. `cu_seq_lens` goes straight to flash-attn's varlen API; without flash-attn, an explicit 4D mask is needed and nothing produces it.

**What I built locally** (drop-in `DataCollator`, ~100 lines):

- FFD bin-packing (identical to existing bfd).
- Emits a 4D `(B, 1, T, T)` attention mask: `0.0` where `same_sample AND j ≤ i`, `-inf` elsewhere.
- Resets `position_ids` to 0 at each sample boundary (RoPE correctness).
- Keeps the 2D pad mask alongside so `compute_loss`'s entropy metric (which assumes 2D) doesn't break.

Verified on real SFT data: per-sample loss matches non-packed to ~0.01 (bf16 deep-network noise floor) on a full 8B model; on the SmolLM2 quick test above the match is ~5e-5 (fp32 noise).

**Since I'm new to contributing to TRL, a few questions before I open anything:**

1. Is this the direction you'd want — keep #5316's whitelist change and add a mask-emitting collator path for non-flash-attn backends — or a different design?
2. Would you prefer a follow-up PR on top of #5316, or bundled in?
3. Pitfalls I should think about ahead of time: `padding_free=True` dataset flag interaction, eval-time packing, `compute_loss` shape expectations?
4. Any convergence test you'd want me to run (e.g., loss-curve equivalence between flash-attn bfd and 4D-mask bfd on a small model)?

Happy to share the collator + test script if you want to eyeball them first before deciding whether a PR makes sense. Thanks for the diagnostic script @qgallouedec — much cleaner methodology than what I had locally.

---

🤖 Drafted with [Claude Code](https://claude.com/claude-code) (Claude Opus 4.7), reviewed and posted by me.
